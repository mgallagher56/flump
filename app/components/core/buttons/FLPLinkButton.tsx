import React from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import { Link } from '@remix-run/react';
import * as pressable from '@zag-js/pressable';
import type { PressEvent } from '@zag-js/pressable/dist/pressable.types';
import { normalizeProps, useMachine } from '@zag-js/react';

interface FLPLinkButtonProps extends ButtonProps {
  preventFocusOnPress?: boolean;
  to: string;
  onPress?: (e: PressEvent) => void;
}

const FLPLinkButton: React.FC<React.PropsWithChildren<FLPLinkButtonProps>> = ({
  children,
  colorScheme = 'blue',
  isDisabled,
  preventFocusOnPress,
  to
}) => {
  const [state, send] = useMachine(
    pressable.machine({
      id: 'pressableBaseButton',
      disabled: isDisabled,
      preventFocusOnPress: preventFocusOnPress ?? true
    })
  );

  const api = pressable.connect(state, send, normalizeProps);

  return (
    <Button colorScheme={colorScheme} as={Link} to={to} variant="link" {...api.pressableProps}>
      {children}
    </Button>
  );
};

export default FLPLinkButton;
