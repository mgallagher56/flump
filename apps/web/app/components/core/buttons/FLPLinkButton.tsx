import { type FC, type PropsWithChildren } from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import { Link } from 'react-router';

interface FLPLinkButtonProps extends ButtonProps {
  to: string;
}

const FLPLinkButton: FC<PropsWithChildren<FLPLinkButtonProps>> = ({
  children,
  colorScheme = 'blue',
  disabled,
  variant = 'link',
  to,
  ...rest
}) => {
  return (
    <Button as={Link} colorScheme={colorScheme} disabled={disabled} to={to} variant={variant} {...rest}>
      {children}
    </Button>
  );
};

export default FLPLinkButton;
