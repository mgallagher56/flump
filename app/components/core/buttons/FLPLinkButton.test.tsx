import React from 'react';

import type { LinkProps } from '@remix-run/react';
import { render } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import FLPLinkButton from './FLPLinkButton';

vi.mock('@remix-run/react', () => ({
  Link: (props: LinkProps): React.ReactNode => <div>{props.children}</div>
}));

describe('FLPLinkButton', () => {
  test('renders as expected', () => {
    const { container, getByText } = render(<FLPLinkButton to="/test">Button</FLPLinkButton>);
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });
  test('renders as expected with a custom class', () => {
    const { container, getByText } = render(
      <FLPLinkButton className="test-class" to="/test">
        Button
      </FLPLinkButton>
    );
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });

  test('renders as expected with a variant', () => {
    const { container, getByText } = render(
      <FLPLinkButton variant="solid" to="/test">
        Button
      </FLPLinkButton>
    );
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });
});
