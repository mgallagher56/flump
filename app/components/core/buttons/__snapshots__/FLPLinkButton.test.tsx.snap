// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FLPLinkButton > renders as expected 1`] = `
<div>
  <div>
    Button
  </div>
</div>
`;

exports[`FLPLinkButton > renders as expected with a custom class 1`] = `
<div>
  <div>
    Button
  </div>
</div>
`;

exports[`FLPLinkButton > renders as expected with a variant 1`] = `
<div>
  <div>
    Button
  </div>
</div>
`;
